kernel/kernel-api/src/main/java/io/delta/kernel/coordinatedcommits/CommitCoordinatorClient.java

```diff
@@ -0,0 +1,168 @@
+import io.delta.kernel.engine.coordinatedcommits.GetCommitsResponse;
+import io.delta.kernel.engine.coordinatedcommits.UpdatedActions;
+import io.delta.kernel.engine.coordinatedcommits.actions.AbstractMetadata;
+import io.delta.kernel.engine.coordinatedcommits.actions.AbstractProtocol;
+import io.delta.kernel.utils.CloseableIterator;
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * The CommitCoordinatorClient is responsible for communicating with the commit coordinator and
+ * backfilling commits. It has four main APIs that need to be implemented:
+ *
+ * <ul>
+ *   <li>{@link #registerTable}: Determine the table config during commit coordinator registration.
+ *   <li>{@link #commit}: Commit a new version of the table.
+ *   <li>{@link #getCommits}: Tracks and returns unbackfilled commits.
+ *   <li>{@link #backfillToVersion}: Ensure that commits are backfilled if/when needed.
+ * </ul>
+ *
+ * @since 3.3.0
+ */
+@Evolving
+public interface CommitCoordinatorClient {
+
+  /**
+   * Register the table represented by the given {@code logPath} at the provided {@code
+   * currentVersion} with the commit coordinator this commit coordinator client represents.
+   *
+   * <p>This API is called when the table is being converted from an existing file system table to a
+   * coordinated-commit table.
+   *
+   * <p>When a new coordinated-commit table is being created, the {@code currentVersion} will be -1
+   * and the upgrade commit needs to be a file system commit which will write the backfilled file
+   * directly.
+   *
+   * @param engine The {@link Engine} instance to use.
+   * @param logPath The path to the delta log of the table that should be converted.
+   * @param tableIdentifier The table identifier for the table, or {@code null} if the table doesn't
+   *     use any identifier (i.e. it is path-based).
+   * @param currentVersion The version of the table just before conversion. currentVersion + 1
+   *     represents the commit that will do the conversion. This must be backfilled atomically.
+   *     currentVersion + 2 represents the first commit after conversion. This will go through the
+   *     CommitCoordinatorClient and the client is free to choose when it wants to backfill this
+   *     commit.
+   * @param currentMetadata The metadata of the table at currentVersion
+   * @param currentProtocol The protocol of the table at currentVersion
+   * @return A map of key-value pairs which is issued by the commit coordinator to uniquely identify
+   *     the table. This should be stored in the table's metadata for table property {@link
+   *     io.delta.kernel.internal.TableConfig#COORDINATED_COMMITS_TABLE_CONF}. This information
+   *     needs to be passed to the {@link #commit}, {@link #getCommits}, and {@link
+   *     #backfillToVersion} APIs to identify the table.
+   */
+  Map<String, String> registerTable(
+      Engine engine,
+      String logPath,
+      @Nullable TableIdentifier tableIdentifier,
```

Comment:
Why don't we use `Optional` like in `TableDescriptor`?

Reply:
I've been flip-flopping on this, myself.

Using Optional is not preferred in a java public API ... but I've been considering refactoring and using Optional actually